kymatio/scattering2d/filter_bank.py

```diff
@@ -36,35 +36,39 @@ def filter_bank(M, N, J, L=8):
     filters = {}
     filters['psi'] = []
 
-    offset_unpad = 0
     for j in range(J):
         for theta in range(L):
             psi = {}
             psi['j'] = j
             psi['theta'] = theta
             psi_signal = morlet_2d(M, N, 0.8 * 2**j,
                 (int(L-L/2-1)-theta) * np.pi / L,
-                3.0 / 4.0 * np.pi /2**j, 4.0/L, offset=offset_unpad)
+                3.0 / 4.0 * np.pi /2**j, 4.0/L)
             psi_signal_fourier = fft2(psi_signal)
+            # drop the imaginary part, it is zero anyway
+            psi_signal_fourier = np.real(psi_signal_fourier)
```

Comment:
Q: Do we ever not want this?

I can't think of a reason other than:

We are inside a function that creates a filter bank by calling a filter function repeatedly.

If we had more than one type of filters, this function would likely be reused, pointing to the fact that either it should be optional to set the imaginary part to 0, or it should already be taken care of by the filters.

Which leads to modifying the filter generator function to ask to be able to have a fourier version of it returned (regardless of how the function generates it)

So we have two options here: Keep this logic as it is, and definitely be required to move it upon refactoring for multiple wavelet types, or push the fft into the filter generation now because we know it needs to happen eventually/soon.

Any thoughts on this? I might be willing to go with the former, but it is pretty clear that the abovementioned refactoring will have to happen pretty soon

Reply:
> I don't understand, are you saying that we'd have morlet_2d generate the (real) Fourier transform directly?

Up to (future) discussion, but yes, essentially that.

The idea would be to delegate the appearance in Fourier to the filter generation function in a kwarg `fourier=True`. If that kwarg isn't present we could also come up with a decorator to add it and make a normal complex fft be returned.
The reason for this is that it seems that the filter function would "know best" how to present its fourier transform, i.e. whether its fourier transform is purely real or not. (We could envision something similar for filters that are purely real in signal domain as well, but first things first)

The filterbank generation function could then just ask the filter function to return its fourier version and not worry about knowing/guessing the properties of the filters.

Reply:
Ok. Taking this idea to its extreme, it seems that we need a `Filter` object. At the end of the day, we don't care about how a filter is represented, only that we can apply it to some array efficiently. This application can be in Fourier (real or complex) or it can be spatial (when we get to that). Depending on the size of the input, we may want to do one or the other. Ultimately, this logic should reside with the filter, not with the `Scattering` objects.